batch_job/src/batch_job_k8s.c

```diff
@@ -0,0 +1,463 @@
+		const struct rmsummary *resources )
+{
+	
+	if(mf_uuid == NULL) {
+		mf_uuid = malloc(sizeof(*mf_uuid));
+		cctools_uuid_create (mf_uuid);
+		// The pod id cannot include upper case
+		string_tolower(mf_uuid->str);
+	}
+	
+	if(k8s_job_info_table == NULL) 
+		k8s_job_info_table = itable_create(0);
+
+	if(k8s_image == NULL) {
+		if((k8s_image = batch_queue_get_option(q, "k8s-image")) == NULL) {
+			fatal("Please specify the container image by using \"--k8s-image\"");
+		}
+	}
+	
+	fflush(NULL);
+	pid_t pid = fork();
+	
+	int job_id = count ++;
+
+	if(pid > 0) {
+
+		debug(D_BATCH, "started job %d: %s", job_id, cmd);
+		struct batch_job_info *info = calloc(1, sizeof(*info));
+		info->submitted = time(0);
+		info->started = time(0);
+		itable_insert(q->job_table, job_id, info);
+		
+		k8s_job_info *curr_job_info = create_k8s_job_info(job_id, cmd,
+				extra_input_files, extra_output_files);
+		
+		itable_insert(k8s_job_info_table, job_id, curr_job_info); 
+		return job_id;
+
+	} else if(pid == 0) {
+
+		if(envlist) {
+			jx_export(envlist);
+		}
+
+		char *pod_id = string_format("%s-%d", mf_uuid->str, job_id);
+		char *k8s_config_fn = string_format("%s.json", pod_id);
+
+		FILE *fd = fopen(k8s_config_fn, "w+");
+		
+		if(resources) {
+			int cores = resources->cores;
+			int memory = resources->memory;
+			char *k8s_cpu, *k8s_memory;
+			if (cores > -1 && memory > -1) {
+				k8s_cpu = string_format("%dm", cores * 1000);
+				k8s_memory = string_format("%dMi", memory);
+			} else {
+				// By default, each container will request
+				// 0.5 cpu and 1024 MB ram
+				k8s_cpu = string_format("%dm", 500);
+				k8s_memory = string_format("%dMi", 1024);
+			}
+			resources_block = string_format(resource_tmpl, 
+					k8s_cpu, k8s_memory, k8s_cpu, k8s_memory);	
+		}
+
+		fprintf(fd, k8s_config_tmpl, mf_uuid->str, pod_id, pod_id, k8s_image, 
+				resources_block, job_id, pod_id);
+
+		fclose(fd);
+
+		if(access(k8s_script_file_name, F_OK | X_OK) == -1) {
+			debug(D_BATCH, "Generating k8s script...");
+			FILE *f = fopen(k8s_script_file_name, "w");
+			fprintf(f, "%s", k8s_script);
+			fclose(f);
+			// Execute permissions
+			chmod(k8s_script_file_name, 0755);
+		}
+
+		char *job_id_str = string_format("%d", job_id);
+
+		execlp("/bin/bash", "bash", k8s_script_file_name, "create", pod_id, job_id_str, 
+				extra_input_files, cmd, extra_output_files, (char *) NULL);
+		_exit(127);
+
+	} else {
+
+		debug(D_BATCH, "couldn't create new process: %s\n", strerror(errno));
+		return -1;
+
+	}
+
+	return -1;
+}
+
+static int batch_job_k8s_remove (struct batch_queue *q, batch_job_id_t jobid)
+{
+	
+	pid_t pid = fork();
+	char *pod_id = string_format("%s-%d", mf_uuid->str, (int)jobid);
+	
+    if(pid > 0) {
+
+    	int status;
+
+    	debug(D_BATCH, "Trying to remove task %d by deleting pods %s.", 
+			(int)jobid, pod_id);
+		
+		// wait child process to complete
+
+		if(waitpid(pid, &status, 0) == -1) {
```

Comment:
I think this needs to use process_waitpid, otherwise we have a garbage collection problem in the process module.